data-prep-pipeline/scripts/subtype_tree/create_subtype_tree.py

```diff
@@ -0,0 +1,619 @@
+            
+            elif lin_node.NodeSource == 'Depmap':
+                ## DECISION: Add at level 1, right underneath custom lineage node
+                parent_code = lin_node.DepmapModelType
+
+        new_nodes.append(
+                construct_new_table_node(
+                    new_code=new_type.DepmapModelType,
+                    new_name=new_type.OncotreeSubtype,
+                    new_source='Depmap',
+                    parent_code=parent_code,
+                    oncotable=oncotable_nc
+                )
+            )
+
+    custom_table = pd.concat([
+        oncotable_nc,
+        pd.DataFrame(new_nodes)
+    ])
+
+    return custom_table
+
+def add_disease_restricted_genetic_subtypes(gs_whitelist, subtype_tree):
+    '''
+    A function to add the white-listed genetic subtypes to the lineage-based tree
+
+    Inputs:
+        - gs_whitelist (pandas df): A table of custom genetic subtypes to add
+        - subtype_tree (pandas df): The subtype tree table
+
+    Outputs:
+        - An extended version of the subtype tree with the additional genetic
+            subtype nodes
+    '''
+    gs_nodes = []
+    for idx, gs in gs_whitelist.iterrows():
+        #create new node
+        gs_node = construct_new_table_node(
+            new_code=gs.MolecularSubtypeCode,
+            new_name=f'{gs.parent_name}: {gs.subtype_name}',
+            new_source='Data-driven genetic subtype',
+            parent_code=gs.parent_code,
+            oncotable=subtype_tree,
+        )
+        gs_nodes.append(gs_node)
+
+    # the code is a data-driven Molecular Subtype, and does not 
+    # originate from gumbo or any other annotation
+    gs_tree = pd.DataFrame(gs_nodes)\
+                .rename(columns={
+                    'DepmapModelType':'MolecularSubtypeCode'
+                })
+    
+    return pd.concat([subtype_tree, gs_tree])
+
+def add_molecular_subtype_subtree(df, mst_tree):
+    '''
+    A function to determine the hierarchy of a gene-specfic subset of the
+    OmicsInferredMolecularSubtype columns. Once the hierarchy is determined,
+    a tree structure is created.
+
+    Inputs:
+        - df (pandas df): a dataframe of the subtypes to add. The function assumes
+            that all subtypes in the df are associated with one gene, and that
+            gene comes at the beginning of each subtype name. Columns in this df
+            are [gene, subtype, full_st], where full_st is the full subtype name
+
+        - mst_tree (pandas df): The molecular subtype tree, which mimics the format 
+            of the subtype tree. In this case, it assumes that all Level0 nodes have
+            been properly added
+
+    Outputs:
+        - mst_tree (pandas df): An extended molecular subtype tree with this gene's
+            sub-tree added in 
+    '''
+    
+    #sorting by full subtype name is important so that the level and parent
+    #assignments don't get jumbled
```

Comment:
I don't think I understand what this comment means. Do you have an example of what the level and parent assignments getting jumbled looks like?

Reply:
Just added a more in-depth explanation of what I mean!